cranelift-codegen/shared/src/isa/x86/encoding_bits.rs

```diff
@@ -27,61 +27,101 @@ use packed_struct::prelude::*;
-    #[packed_field(bits = "8:11", ty = "enum")]
-    pub prefix: OpcodePrefix,
-
-    /// Bits for the ModR/M byte for certain opcodes.
-    #[packed_field(bits = "12:14")]
-    pub rrr: Integer<u8, packed_bits::Bits3>,
-
-    /// REX.W bit (or VEX.W/E).
-    #[packed_field(bits = "15")]
-    pub rex_w: Integer<u8, packed_bits::Bits1>,
-}
+#[derive(Copy, Clone, PartialEq)]
+pub struct EncodingBits(u16);
+const OPCODE: RangeInclusive<u16> = 0..=7;
+const OPCODE_PREFIX: RangeInclusive<u16> = 8..=11; // Includes pp and mm.
+const RRR: RangeInclusive<u16> = 12..=14;
+const REX_W: RangeInclusive<u16> = 15..=15;
 
 impl From<u16> for EncodingBits {
-    fn from(bits: u16) -> EncodingBits {
-        let bytes: [u8; 2] = [((bits >> 8) & 0xff) as u8, (bits & 0xff) as u8];
-        EncodingBits::unpack(&bytes).expect("failed creating EncodingBits")
+    fn from(bits: u16) -> Self {
+        Self(bits)
     }
 }
 
 impl EncodingBits {
     /// Constructs a new EncodingBits from parts.
     pub fn new(op_bytes: &[u8], rrr: u16, rex_w: u16) -> Self {
-        EncodingBits {
-            opcode_byte: op_bytes[op_bytes.len() - 1],
-            prefix: OpcodePrefix::from_opcode(op_bytes),
-            rrr: (rrr as u8).into(),
-            rex_w: (rex_w as u8).into(),
-        }
+        assert!(
+            op_bytes.len() > 0,
+            "op_bytes must at least include an opcode byte"
+        );
+        let mut new = Self::from(0);
+        let last_byte = op_bytes[op_bytes.len() - 1];
+        new.write(OPCODE, last_byte as u16);
+        let prefix: u8 = OpcodePrefix::from_opcode(op_bytes).into();
+        new.write(OPCODE_PREFIX, prefix as u16);
+        new.write(RRR, rrr);
+        new.write(REX_W, rex_w);
+        new
     }
 
     /// Returns the raw bits.
     #[inline]
     pub fn bits(self) -> u16 {
-        let bytes: [u8; 2] = self.pack();
-        ((bytes[0] as u16) << 8) | (bytes[1] as u16)
+        self.0
+    }
+
+    /// Convenience method for writing bits to specific range.
+    #[inline]
+    fn write(&mut self, range: RangeInclusive<u16>, value: u16) {
+        let size = range.end() - range.start() + 1; // Calculate the number of bits in the range.
+        let mask = (1 << size) - 1; // Generate a bit mask.
+        debug_assert!(
+            value <= mask,
+            "The written value should have fewer than {} bits.",
+            size
+        );
+        let value = (value & mask) << *range.start(); // Place the value in the correct locations.
+        let complement = !value; // Create the bitwise complement for the clear mask.
+        self.0 &= complement; // Clear the bits in `range`.
+        self.0 |= value; // Modify the bits in `range`.
+    }
+
+    /// Convenience method for reading bits from a specific range.
+    #[inline]
+    fn read(self, range: RangeInclusive<u16>) -> u8 {
+        let size = range.end() - range.start() + 1; // Calculate the number of bits in the range.
```

Comment:
The `RangeInclusive<u16>` doesn't actually enforce that `start <= end`. It merely guarantees that if given a range with an empty inclusive range, as an iterator it will not yield any elements.

However, if given such a range, the documentation for `RangeInclusive::start()` and `RangeInclusive::end()` say that calling `start()` and `end()` yield *unspecified values*. The documentation says that before using those function, the existence of a valid range must be checked explicitly via `is_empty()`.

Reply:
Maybe it's a Rust version issue or something (I'm on 1.39.0) but when I add the assert I see the following error:

```
error[E0034]: multiple applicable items in scope
  --> cranelift-codegen/shared/src/isa/x86/encoding_bits.rs:70:24
   |
70 |         assert!(!range.is_empty());
   |                        ^^^^^^^^ multiple `is_empty` found
   |
   = note: candidate #1 is defined in an impl for the type `std::ops::RangeInclusive<_>`
   = note: candidate #2 is defined in an impl of the trait `std::iter::ExactSizeIterator` for the type `std::ops::RangeInclusive<u16>`
   = help: to disambiguate the method call, write `std::iter::ExactSizeIterator::is_empty(range)` instead
```

When I take the compiler's advice I still see the error and adding feature flags as in https://github.com/rust-lang/rust/issues/48111#issuecomment-445132965 does not seem to work either. I may have to create a new issue for this...

Reply:
Opened https://github.com/bytecodealliance/cranelift/issues/1284.

Reply:
Both `is_empty` methods are still unstable. Maybe use `range.len()` instead of calculating the size yourself?

Reply:
Does it have `len`? I can't seem to find it...

Reply:
I haven't tried, but it should have it via the `ExactSizeIterator` impl.

Reply:
Thanks! I think that works.